manual/src/tutorials/gadtexamples.etex

```diff
@@ -309,3 +309,119 @@ contradict the exhaustiveness check. Namely, builtin types (those
 defined by the compiler itself, such as "int" or "array"), and
 abstract types defined by the local module, are non-instantiable, and
 as such cause a type error rather than introduce an equation.
+
+
+\section{s:types-used-as-gadt-indices}{Types used as GADT indices}
+
+In our discussion of refutation cases (\ref{s:gadt-refutation-cases}), we used the following example:
+
+\begin{caml_example*}{verbatim}
+type _ t =
+  | Int : int t
+  | Bool : bool t
+
+let deep : (char t * int) option -> char = function
+  | None -> 'c'
+  | _ -> .
+\end{caml_example*}
+
+The type-checker is able to verify that the "Some _" case is
+impossible becaus the type "char t" is empty. This relies on the fact
+that the types occurring in the GADT index are known to be distinct:
+the constructor "Int" returns an "int t", and we know that "int" and
+"char" are distinct types, so "Int" is impossible at type "char t" --
+and similarly for "Bool". In other words, checking that a GADT case is
+impossible (at a specific type) requires reasoning on inequalities
+between types.
+
+For this reason, using abstract types as GADT indices does not work as
+intended. Consider for instance:
+
+\begin{caml_example*}{verbatim}
+(* type-level natural numbers, in unary notation *)
+module TypeNat = struct
+  type zero
+  type !'a succ
+end
+
+type ('a, _) sized_list =
+  | Nil : ('a, TypeNat.zero) sized_list
+  | Cons : 'a * ('a, 'n) sized_list -> ('a, 'n TypeNat.succ) sized_list
+
+type one = TypeNat.(zero succ)
+\end{caml_example*}
+
+This definition does not work: because "zero" and "succ" are abstract
+types, the compiler assumes that their definition is unknown, and that
+they might in fact be the same type. In particular, the types "zero"
+and "zero succ" are not known to be distinct, and the compiler cannot
+agree that the "Nil" case is impossible at type "zero
+succ":\footnote{We also needed an explicit injectivity annotation,
+  "type !'a succ", otherwise the definition of "sized_list" is also
+  rejected because "succ" is abstract.}
+\begin{caml_example}{verbatim}[error]
+let of_singleton : ('a, one) sized_list -> 'a = function
+  | Nil -> .
+  | Cons (v, Nil) -> v
+  | Cons (v, Cons _) -> .
+\end{caml_example}
+
+The recommended way to avoid this problem is to give a concrete
+definition to types such as "zero" and "succ", that are only meant to
+be used in GADT indices. (This is counter-intuitive, as we will never
+manipulate values of this type.) You could use:
+\begin{caml_example*}{verbatim}
+module TypeNat = struct
+  type zero = Zero
+  type 'a succ = Succ
+end
+\end{caml_example*}
+
+As an additional refinement, we can define those index types using
+using a "private" annotation, which clarifies that they are not meant
+to be used to build values:
+\begin{caml_example*}{verbatim}
+module TypeNat = struct
+  type zero = private Zero
+  type 'a succ = private Succ
```

Comment:
Note: the idiom that I find online uses polymorphic variants, ``type zero = private [ `Zero ]``, but at the time of writing this I could not reconstruct a reason for using polymorphic variants rather than normal variants (`type zero = private Zero`). Why would we use polymorphic variants here?

Reply:
I see no specific reason. At least, from the point of view of `Ctype.mcomp`, using normal or polymorphic variants are strictly equivalent. Namely, they are both structural.
There are still some differences in usage:
* `type zero = private Zero` defines a constructor (so it may be interpreted as namespace pollution) and while it prevents using `Zero` in expression, one can still use it in patterns.
* `type zero = private [``Zero]` is actually closer to `type zero = external "zero"` as it has no impact on the environment. If it is the goal, then `external` seems ultimately better.

Reply:
Adding things to the environment is a good point. On the other hand, the non-polymorphic-variant version involves simpler language features and is easier on the eyes.